xtm/xtm_scsp_queue.h

```diff
@@ -0,0 +1,163 @@
+ * copying operation is carried out quickly and correctly.
+ */
+template<class T>
+struct xtm_scsp_queue {
+	/*
+	 * Next position to be written
+	 */
+	unsigned write;
+	/*
+	 * Next position to be read
+	 */
+	unsigned read;
+	/*
+	 * Circular buffer length
+	 */
+	unsigned len;
+	/*
+	 * Buffer contains objects
+	 */
+	T buffer[];
+};
+
+/**
+ * Function that initializes queue.
+ * @param[out] queue - queue of elements with type T
+ * @param[in] size - size of queue, must be power of 2
+ * @retval return 0 if success otherwise return -1
+ */
+template <class T>
+static inline int
+xtm_scsp_queue_init(struct xtm_scsp_queue<T> *queue, unsigned size)
+{
+	/* Ensure size is power of 2 */
+	if (size & (size - 1))
+		return -1;
+
+	queue->write = 0;
+	queue->read = 0;
+	queue->len = size;
+	return 0;
+}
+
+/**
+ * Function adds num elements into the queue.
+ * @param[in] queue - queue of elements with type T
+ * param[in] data - array of elements to add
+ * param[in] num - count of elements to add
+ * @retval return the number of elements actually written.
+ */
+template <class T>
+static inline unsigned
+xtm_scsp_queue_put(struct xtm_scsp_queue<T> *queue, T *data, unsigned num)
+{
+	unsigned i = 0;
+	unsigned queue_write = queue->write;
+	unsigned new_write = queue_write;
+	unsigned queue_read =
+		__atomic_load_n(&queue->read, __ATOMIC_ACQUIRE);
+
+	for (i = 0; i < num; i++) {
+		new_write = (new_write + 1) & (queue->len - 1);
+		if (new_write == queue_read)
+			break;
+		queue->buffer[queue_write] = data[i];
+		queue_write = new_write;
+	}
+	__atomic_store_n(&queue->write, queue_write, __ATOMIC_RELEASE);
+	return i;
+}
+
+/**
+ * Function gets up to num elements from the queue.
+ * @param[in] queue - queue of elements with type T
+ * @param[out] data - array to save elements
+ * @param[in] num - count of elements to get
+ * @retval return the number of elements actually read.
+ */
+template <class T>
+static inline unsigned
+xtm_scsp_queue_get(struct xtm_scsp_queue<T> *queue, T *data, unsigned num)
+{
+	unsigned i = 0;
+	unsigned new_read = queue->read;
+	unsigned queue_write =
+		__atomic_load_n(&queue->write, __ATOMIC_ACQUIRE);
+
+	for (i = 0; i < num; i++) {
+		if (new_read == queue_write)
+			break;
+		data[i] = queue->buffer[new_read];
+		new_read = (new_read + 1) & (queue->len - 1);
+	}
+	__atomic_store_n(&queue->read, new_read, __ATOMIC_RELEASE);
+	return i;
+}
+
+/**
+ * Function get the num of available elements in the queue
+ * @param[in] queue - queue of elements with type T
+ * @retval return num of available elements in the queue
+ */
+template <class T>
+static inline unsigned
+xtm_scsp_queue_free_count(struct xtm_scsp_queue<T> *queue)
+{
+	unsigned queue_write =
+		__atomic_load_n(&queue->write, __ATOMIC_ACQUIRE);
+	unsigned queue_read =
+		__atomic_load_n(&queue->read, __ATOMIC_ACQUIRE);
+	return (queue_read - queue_write - 1) & (queue->len - 1);
+}
+
+/**
+ * Function get the num of elements in the queue
```

Comment:
You don't need to write 'Function' in each comment. Function comments must be in imperative mood like the commit titles saying what the function should do.
The same for all the other places in all the files.

Reply:
Fixed